obstore/python/obstore/auth/azure.py

```diff
@@ -0,0 +1,200 @@
+
+from datetime import UTC, datetime, timedelta
+from typing import TYPE_CHECKING
+
+import azure.identity
+import azure.identity.aio
+
+if TYPE_CHECKING:
+    from obstore.store import AzureCredential
+
+    AzureCredentialUnionType = (
+        azure.identity.AuthorizationCodeCredential
+        | azure.identity.AzureCliCredential
+        | azure.identity.AzureDeveloperCliCredential
+        | azure.identity.AzurePipelinesCredential
+        | azure.identity.AzurePowerShellCredential
+        | azure.identity.CertificateCredential
+        | azure.identity.ChainedTokenCredential
+        | azure.identity.ClientAssertionCredential
+        | azure.identity.ClientSecretCredential
+        | azure.identity.DefaultAzureCredential
+        | azure.identity.DeviceCodeCredential
+        | azure.identity.EnvironmentCredential
+        | azure.identity.InteractiveBrowserCredential
+        | azure.identity.ManagedIdentityCredential
+        | azure.identity.OnBehalfOfCredential
+        | azure.identity.SharedTokenCacheCredential
+        | azure.identity.UsernamePasswordCredential
+        | azure.identity.VisualStudioCodeCredential
+        | azure.identity.WorkloadIdentityCredential
+    )
+
+    AzureAsyncCredentialUnionType = (
+        azure.identity.aio.AuthorizationCodeCredential
+        | azure.identity.aio.AzureCliCredential
+        | azure.identity.aio.AzureDeveloperCliCredential
+        | azure.identity.aio.AzurePipelinesCredential
+        | azure.identity.aio.AzurePowerShellCredential
+        | azure.identity.aio.CertificateCredential
+        | azure.identity.aio.ChainedTokenCredential
+        | azure.identity.aio.ClientAssertionCredential
+        | azure.identity.aio.ClientSecretCredential
+        | azure.identity.aio.DefaultAzureCredential
+        | azure.identity.aio.EnvironmentCredential
+        | azure.identity.aio.ManagedIdentityCredential
+        | azure.identity.aio.OnBehalfOfCredential
+        | azure.identity.aio.SharedTokenCacheCredential
+        | azure.identity.aio.VisualStudioCodeCredential
+        | azure.identity.aio.WorkloadIdentityCredential
+    )
+
+
+class AzureAuthCredentialProvider:
+    """A CredentialProvider for [AzureStore][obstore.store.AzureStore] that uses [`azure.identity`][].
+
+    This credential provider uses `azure-identity`, and will error if this cannot
+    be imported.
+
+    **Example:**
+
+    ```py
+    from obstore.auth.azure import AzureAuthCredentialProvider
+    from obstore.store import AzureStore
+
+    credential_provider = AzureAuthCredentialProvider(credential=...)
+    store = AzureStore("container", credential_provider=credential_provider)
+    ```
+
+    [`azure.identity`]: https://learn.microsoft.com/en-us/python/api/overview/azure/identity-readme
+    """  # noqa: E501
+
+    credential: AzureCredentialUnionType
+
+    def __init__(
+        self,
+        credential: AzureCredentialUnionType | None = None,
+        scopes: list[str] | None = None,
```

Comment:
And then here, this can be
```suggestion
        scopes: Iterable[str] = DEFAULT_SCOPES,
```

Reply:
Done.